File: src/main/java/com/uwetrottmann/trakt5/entities/EpisodeCheckin.java
/*
 * Copyright 2024 Uwe Trottmann
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.uwetrottmann.trakt5.entities;

import javax.annotation.Nonnull;

public class EpisodeCheckin extends BaseCheckin {

    public Show show;
    public SyncEpisode episode;

    public static class Builder {

        private Show show;
        private SyncEpisode episode;
        protected ShareSettings sharing;
        protected String message;
        protected String venue_id;
        protected String venue_name;
        protected String app_version;
        protected String app_date;

        public Builder(SyncEpisode episode, String appVersion, String appDate) {
            if (episode == null) {
                throw new IllegalArgumentException("Episode must not be null");
            }
            this.episode = episode;
            this.app_version = appVersion;
            this.app_date = appDate;
        }

        @Nonnull
        public Builder show(Show show) {
            this.show = show;
            return this;
        }

        @Nonnull
        public Builder sharing(ShareSettings shareSettings) {
            this.sharing = shareSettings;
            return this;
        }

        @Nonnull
        public Builder message(String message) {
            this.message = message;
            return this;
        }

        @Nonnull
        public Builder venueId(String venueId) {
            this.venue_id = venueId;
            return this;
        }

        @Nonnull
        public Builder venueName(String venueName) {
            this.venue_name = venueName;
            return this;
        }

        @Nonnull
        public EpisodeCheckin build() {
            EpisodeCheckin checkin = new EpisodeCheckin();
            checkin.show = show;
            checkin.episode = episode;
            checkin.sharing = sharing;
            checkin.message = message;
            checkin.venue_id = venue_id;
            checkin.venue_name = venue_name;
            checkin.app_date = app_date;
            checkin.app_version = app_version;
            return checkin;
        }
    }

}

File: src/main/kotlin/com/redissi/trakt/entities/EpisodeCheckin.kt
package com.redissi.trakt.entities

import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass

@JsonClass(generateAdapter = true)
class EpisodeCheckin(
    override val sharing: ShareSettings? = null,
    override val message: String? = null,
    @Json(name = "venue_id")
    override val venueId: String? = null,
    @Json(name = "venue_name")
    override val venueName: String? = null,
    @Json(name = "app_version")
    override val appVersion: String? = null,
    @Json(name = "app_date")
    override val appDate: String? = null,
    val show: Show? = null,
    val episode: SyncEpisode? = null
) : BaseCheckin() {

    class Builder(
        episode: SyncEpisode?,
        appVersion: String,
        appDate: String
    ) {
        private var show: Show? = null
        private val episode: SyncEpisode
        private var sharing: ShareSettings? = null
        private var message: String? = null
        private var venueId: String? = null
        private var venueName: String? = null
        private var appVersion: String
        private var appDate: String
        fun show(show: Show?): Builder {
            this.show = show
            return this
        }

        fun sharing(shareSettings: ShareSettings?): Builder {
            sharing = shareSettings
            return this
        }

        fun message(message: String?): Builder {
            this.message = message
            return this
        }

        fun venueId(venueId: String?): Builder {
            this.venueId = venueId
            return this
        }

        fun venueName(venueName: String?): Builder {
            this.venueName = venueName
            return this
        }

        fun build(): EpisodeCheckin {
            return EpisodeCheckin(
                show = show,
                episode = episode,
                sharing = sharing,
                message = message,
                venueId = venueId,
                venueName = venueName,
                appDate = appDate,
                appVersion = appVersion
            )
        }

        init {
            requireNotNull(episode) { "Episode must not be null" }
            this.episode = episode
            this.appVersion = appVersion
            this.appDate = appDate
        }
    }
}